File: app_manager/appmanager/src/main/java/app/editors/manager/ui/adapters/holders/CloudAccountsViewHolder.java
package app.editors.manager.ui.adapters.holders;

import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageButton;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import app.editors.manager.R;
import app.editors.manager.managers.utils.GlideUtils;
import app.editors.manager.managers.utils.UiUtils;
import app.editors.manager.mvp.models.account.AccountsSqlData;
import app.editors.manager.ui.adapters.CloudAccountsAdapter;
import butterknife.BindView;
import butterknife.ButterKnife;

public class CloudAccountsViewHolder extends RecyclerView.ViewHolder {

    @BindView(R.id.accountItemLayout)
    public ConstraintLayout mAccountLayout;
    @BindView(R.id.view_icon_selectable_image)
    AppCompatImageView mViewIconSelectableImage;
    @BindView(R.id.imageCheck)
    AppCompatImageView mViewCheckImage;
    @BindView(R.id.view_icon_selectable_mask)
    FrameLayout mViewIconSelectableMask;
    @BindView(R.id.selectableLayout)
    FrameLayout mViewIconSelectableLayout;
    @BindView(R.id.accountItemName)
    AppCompatTextView mAccountName;
    @BindView(R.id.accountItemPortal)
    AppCompatTextView mAccountPortal;
    @BindView(R.id.accountItemEmail)
    AppCompatTextView mAccountEmail;
    @BindView(R.id.accountItemContext)
    public AppCompatImageButton mAccountContext;

    public CloudAccountsViewHolder(@NonNull View itemView) {
        super(itemView);
        ButterKnife.bind(this, itemView);
    }

    public void bind(AccountsSqlData account,
                     boolean isSelection,
                     @Nullable CloudAccountsAdapter.OnAccountClick accountClick,
                     @Nullable CloudAccountsAdapter.OnAccountLongClick accountLongClick,
                     @Nullable CloudAccountsAdapter.OnAccountContextClick accountContextClick) {
        mAccountName.setText(account.getName());
        mAccountPortal.setText(account.getPortal());
        mAccountEmail.setText(account.getLogin());

        mAccountContext.setVisibility(View.VISIBLE);

        if (!isSelection) {
            if (account.isOnline()) {
                mViewCheckImage.setVisibility(View.VISIBLE);
            } else {
                mViewCheckImage.setVisibility(View.GONE);
            }
        } else {
            mViewCheckImage.setVisibility(View.GONE);
        }

        if (account.isWebDav()) {
            mAccountName.setVisibility(View.GONE);
            UiUtils.setWebDavImage(account.getWebDavProvider(), mViewIconSelectableImage);
        } else {
            mAccountName.setVisibility(View.VISIBLE);
            final String url = account.getScheme() + account.getPortal() + account.getAvatarUrl();
            Glide.with(mViewIconSelectableImage)
                    .load(GlideUtils.getCorrectLoad(url, account.getToken()))
                    .apply(GlideUtils.getAvatarOptions())
                    .into(mViewIconSelectableImage);
        }

        mViewIconSelectableLayout.setBackground(null);
        mViewIconSelectableMask.setBackground(null);
        if (isSelection) {
            mAccountContext.setVisibility(View.GONE);
            if (account.isSelection()) {
                mViewIconSelectableMask.setBackgroundResource(R.drawable.drawable_list_image_select_mask);
            } else {
                mViewIconSelectableMask.setBackgroundResource(R.drawable.drawable_list_image_select_background);
            }
        }

        setListener(account, accountClick, accountLongClick, accountContextClick);
    }

    private void setListener(AccountsSqlData account,
                             CloudAccountsAdapter.OnAccountClick accountClick,
                             CloudAccountsAdapter.OnAccountLongClick accountLongClick,
                             CloudAccountsAdapter.OnAccountContextClick accountContextClick) {
        mAccountLayout.setOnClickListener(v -> {
            if (accountClick != null) {
                accountClick.onAccountClick(account, getLayoutPosition());
            }
        });

        mAccountLayout.setOnLongClickListener(v -> {
            if (accountLongClick != null) {
                accountLongClick.onAccountLongClick(account, getLayoutPosition());
                return true;
            } else {
                return false;
            }
        });

        mAccountContext.setOnClickListener(v -> {
            if (accountContextClick != null) {
                accountContextClick.onAccountContextClick(account, getLayoutPosition(), v);
            }
        });

    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/adapters/holders/CloudAccountsViewHolder.kt
package app.editors.manager.ui.adapters.holders

import android.view.View
import android.widget.FrameLayout
import androidx.appcompat.widget.AppCompatImageButton
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import app.documents.core.account.CloudAccount
import app.editors.manager.R
import app.editors.manager.managers.utils.GlideUtils
import app.editors.manager.managers.utils.ManagerUiUtils
import com.bumptech.glide.Glide

class CloudAccountsViewHolder(view: View) : RecyclerView.ViewHolder(view) {

    private val mAccountLayout: ConstraintLayout = view.findViewById(R.id.accountItemLayout)
    private val mViewIconSelectableImage: AppCompatImageView = view.findViewById(R.id.view_icon_selectable_image)
    private val mViewCheckImage: AppCompatImageView = view.findViewById(R.id.imageCheck)
    private val mViewIconSelectableMask: FrameLayout = view.findViewById(R.id.view_icon_selectable_mask)
    private val mViewIconSelectableLayout: FrameLayout = view.findViewById(R.id.selectableLayout)
    private val mAccountName: AppCompatTextView = view.findViewById(R.id.accountItemName)
    private val mAccountPortal: AppCompatTextView = view.findViewById(R.id.accountItemPortal)
    private val mAccountContext: AppCompatImageButton = view.findViewById(R.id.accountItemContext)

    fun bind(
        account: CloudAccount,
        isSelection: Boolean,
        accountClick: ((account: CloudAccount) -> Unit)? = null,
        accountLongClick: ((account: CloudAccount) -> Unit)? = null,
        accountContextClick: ((account: CloudAccount, position: Int, view: View) -> Unit)? = null
    ) {
        mAccountName.text = account.name
        mAccountPortal.text = account.portal
        mAccountContext.visibility = View.VISIBLE
        if (!isSelection) {
            if (account.isOnline) {
                mViewCheckImage.visibility = View.VISIBLE
            } else {
                mViewCheckImage.visibility = View.GONE
            }
        } else {
            mViewCheckImage.visibility = View.GONE
        }
        if (account.isWebDav) {
            mAccountName.visibility = View.GONE
            ManagerUiUtils.setWebDavImage(account.webDavProvider, mViewIconSelectableImage)
        } else {
            mAccountName.visibility = View.VISIBLE
            account.token?.let { token ->
                val url: String = if (account.avatarUrl?.contains("static") == true) {
                    account.avatarUrl ?: ""
                } else {
                    account.scheme + account.portal + account.avatarUrl
                }
                Glide.with(mViewIconSelectableImage)
                    .load(GlideUtils.getCorrectLoad(url, token))
                    .apply(GlideUtils.avatarOptions)
                    .into(mViewIconSelectableImage)
            }

        }
        mViewIconSelectableLayout.background = null
        mViewIconSelectableMask.background = null
        if (isSelection) {
            mAccountContext.visibility = View.GONE
//            if (account.isSelection) {
//                mViewIconSelectableMask.setBackgroundResource(R.drawable.drawable_list_image_select_mask)
//            } else {
//                mViewIconSelectableMask.setBackgroundResource(R.drawable.drawable_list_image_select_background)
//            }
        }
        setListener(account, accountClick, accountLongClick, accountContextClick)
    }

    private fun setListener(
        account: CloudAccount,
        accountClick: ((account: CloudAccount) -> Unit)? = null,
        accountLongClick: ((account: CloudAccount) -> Unit)? = null,
        accountContextClick: ((account: CloudAccount, position: Int, view: View) -> Unit)? = null
    ) {
        mAccountLayout.setOnClickListener {
            accountClick?.invoke(account)
        }
        mAccountLayout.setOnLongClickListener {
            accountLongClick?.let { listener ->
                listener.invoke(account)
                return@setOnLongClickListener true
            } ?: run {
                return@setOnLongClickListener false
            }

        }
        mAccountContext.setOnClickListener { view: View ->
            accountContextClick?.invoke(account, absoluteAdapterPosition, view)
        }
    }

}